contracts/GnosisSafePersonalEdition.sol

```diff
@@ -0,0 +1,132 @@
+/// @author Richard Meissner - <richard@gnosis.pm>
+contract GnosisSafePersonalEdition is MasterCopy, GnosisSafe {
+
+    string public constant NAME = "Gnosis Safe Personal Edition";
+    string public constant VERSION = "0.0.1";
+    
+    uint256 internal constant BASE_TX_GAS_COSTS = 21000;
+    uint256 internal constant PAYMENT_GAS_COSTS = 10000;
+    // The maximum price for data payload is 68 (non-zero byte), but the avera is way lower, therefore we use 40
+    uint256 internal constant DATA_BYTE_GAS_COSTS = 40;
+
+    event ExecutionFailed();
+
+    uint256 public nonce;
+
+    /// @dev Allows to execute a Safe transaction confirmed by required number of owners.
+    /// @param to Destination address of Safe transaction.
+    /// @param value Ether value of Safe transaction.
+    /// @param data Data payload of Safe transaction.
+    /// @param operation Operation type of Safe transaction.
+    /// @param safeTxGas Gas that should be used for the Safe transaction.
+    /// @param gasPrice Gas price that should be used for the payment calculation.
+    /// @param v Array of signature V values sorted by owner addresses.
+    /// @param r Array of signature R values sorted by owner addresses.
+    /// @param s Array of signature S values sorted by owner addresses.
+    function payAndExecuteTransaction(
+        address to, 
+        uint256 value, 
+        bytes data, 
+        Enum.Operation operation, 
+        uint256 safeTxGas,
+        uint256 gasPrice,
+        uint8[] v, 
+        bytes32[] r, 
+        bytes32[] s
+    )
+        public
+    {
+        uint256 startGas = gasleft();
+        require(address(this).balance - value >= totalGasCosts(safeTxGas) * gasPrice);
+        checkHash(getTransactionHash(to, value, data, operation, safeTxGas, gasPrice, nonce), v, r, s);
+        // Increase nonce and execute transaction.
+        nonce++;
+        require(gasleft() - PAYMENT_GAS_COSTS >= safeTxGas);
+        if (!execute(to, value, data, operation, safeTxGas)) {
+            emit ExecutionFailed();
+        }
+        uint256 gasCosts = totalGasCosts(startGas - gasleft());
+
+        // We transfer the calculated tx costs to the tx.origin to avoid sending it to intermediate contracts that have made calls
+        // solium-disable-next-line security/no-tx-origin
+        tx.origin.transfer(gasCosts * gasPrice);
+    }
+
+    /// @dev Calculates the total gas costs for a safe transaction with the gas costs for the execution of the transaction.
+    /// @param executionGas Gas costs for the execution of the safe transaction.
+    /// @return Total gas costs for the execution (this includes gas costs for the payment to tx.origin, base transaction and payload data).
+    function totalGasCosts(uint256 executionGas) 
+        public 
+        pure
+        returns (uint256) 
+    {
+        return executionGas + PAYMENT_GAS_COSTS + BASE_TX_GAS_COSTS + msg.data.length * DATA_BYTE_GAS_COSTS;
```

Comment:
The tx service has the possibility to add a lot of useless payload to the transaction, to let the user pay more than necessary. Maybe we should include the data.length in the signature.

Reply:
In this case we maybe should precalculate the whole data costs

Reply:
Agreed, this would make sense.